proposals/NNNN-experiment-tracking/README.md

```diff
@@ -0,0 +1,510 @@
+
+### SDK Implementation Details
+
+#### MLFlow Model Registry Plugin
+
+The MLFlow SDK has achieved remarkable adoption in the machine learning community, with over
+[747 million downloads](https://pepy.tech/projects/mlflow?timeRange=threeMonths) on PyPI as of this writing. This
+widespread popularity makes MLFlow SDK compatibility essential for Kubeflow's experiment tracking solution to gain
+traction among the AI/ML community. The MLFlow SDK's extensible
+[plugin architecture](https://mlflow.org/docs/latest/ml/plugins/#storage-plugins) allows developers to swap out the
+underlying backend storage while maintaining the familiar MLFlow API surface.
+
+To enable seamless integration, we will implement two key MLFlow plugins that connect to the Model Registry backend:
+
+1. **Tracking Store Plugin**: This plugin will handle experiment and run metadata, metrics, parameters, and tags. It
+   will map MLFlow's tracking concepts to the Model Registry's domain models (Experiments, Runs, Users, Metrics, etc.)
+   while excluding tracing functionality for the initial implementation.
+
+2. **Artifact Repository Plugin**: This plugin will manage artifact storage and retrieval, leveraging the Model
+   Registry's artifact management capabilities to store model files, datasets, and other ML artifacts.
+
+With these plugins in place, users will be able to use the MLFlow SDK with minimal changes to their existing code. They
+simply need to install an additional Python package and specify the Model Registry API as their tracking URI. This
+approach enables users to leverage MLFlow's powerful
+[automatic logging](https://mlflow.org/docs/latest/ml/tracking/autolog/) functionality for popular ML frameworks
+(TensorFlow, PyTorch, scikit-learn, etc.) while storing all data in the centralized Kubeflow Model Registry.
+
+**Future Enhancements**: In subsequent phases, we may extend the plugin architecture to include:
+
+- **Tracing Support**: Adding tracing capabilities akin to [MLflow](https://mlflow.org/docs/latest/genai/tracing/) would
+  require new domain models in Model Registry.
+- **Model Registry Store**: Further development could expose the Model Registry model registration functionality in the
+  MLFlow SDK.
+
+The primary objective remains enabling users to leverage MLFlow's mature ecosystem and familiar APIs while benefiting
+from Kubeflow's centralized experiment tracking infrastructure.
+
+#### Kubeflow SDK Implementation
+
+Building on the MLFlow Model Registry plugin, we can enhance the Kubeflow SDK to provide a native Kubeflow-centric
+experience. The initial implementation would focus on simplifying setup by automatically configuring the MLFlow Model
+Registry plugin and creating convenient wrapper functions around the supported MLFlow SDK APIs. This integration should
+feel cohesive with the existing Model Registry SDK functionality for registering models.
+
+Enhancements could include:
```

Comment:
I'm trying to understand the distinction between what the Kubeflow SDK should provide versus what the model registry SDK will provide. My understand is that the MR SDK will contain the mlflow plugin, plus an updated client sdk to reflect the new REST APIs. The 4 points below for me would more naturally live in the model registry SDK as they are all Registry/Experiment specific features.
I think specifying the scope of features for the Model Registry SDK will allow us to know what to build upon in the Kubeflow SDK. I think this will avoid duplication of effort.

Reply:
> I'm trying to understand the distinction between what the Kubeflow SDK should provide versus what the model registry SDK will provide. My understand is that the MR SDK will contain the mlflow plugin, plus an updated client sdk to reflect the new REST APIs. 

The MLFlow plugin won't be part of the MR SDK per se, it's just a module users can install to integrate with MR using MLflow SDK. 

The MR SDK will have it's own APIs for Experimentation as well as the existing RegisteredModel resources. So, I agree that duplicating Experimentation support in a separate Kubeflow SDK doesn't add value. Kubeflow SDK could instead focus on making it easier to locate MR services and configure an MR SDK client that users could use  to access MR API.

Reply:
@dhirajsb when you say "The MR SDK will have it's own APIs for Experimentation", are you saying the OpenAPI  generated client code or custom methods like `log_metric`?

If it's the OpenAPI generated code, my thought is that the Kubeflow SDK can be the integration point for Model Registry and the Model Registry MLFlow plugin like the POC. We later create bespoke methods like `log_metric` in the Model Registry SDK and replace those transparently in the Kubeflow SDK if we need more functionality.